teuthology/task/__init__.py

```diff
@@ -0,0 +1,116 @@
+import logging
+
+from teuthology.misc import deep_merge
+
+log = logging.getLogger(__name__)
+
+
+class Task(object):
+    """
+    A base-class for "new-style" teuthology tasks.
+
+    Can be used as a drop-in replacement for the old-style task functions with
+    @contextmanager decorators.
+    """
+    def __init__(self, ctx=None, config=None):
+        self.log = log
+        self.name = self.__class__.__name__.lower()
+        self.ctx = ctx
+        self.config = config or dict()
+        if not isinstance(self.config, dict):
+            raise TypeError("config must be a dict")
+
+    def apply_overrides(self):
+        """
+        Look for an 'overrides' dict in self.ctx.config; look inside that for a
+        dict with the same name as this task. Override any settings in
+        self.config with those overrides
+        """
+        all_overrides = self.ctx.config.get('overrides', dict())
+        if not all_overrides:
+            return
+        task_overrides = all_overrides.get(self.name)
+        if task_overrides:
+            self.log.debug(
+                "Applying overrides for task {name}: {overrides}".format(
+                    name=self.name, overrides=task_overrides)
+            )
+            deep_merge(self.config, task_overrides)
+
+    def find_hosts(self):
+        """
+        Look for a 'hosts' list in self.config. Each item in the list may
+        either be a role or a hostname. The task will then be run against only
+        those hosts which match one (or more) of the roles or hostnames
+        specified.
+        """
+        if not hasattr(self.ctx, 'cluster'):
+            return
+        host_specs = self.config.get('hosts', list())
+        hosts = set()
+        for host_spec in host_specs:
+            role_matches = self.ctx.cluster.only(host_spec)
+            if len(role_matches.remotes) > 0:
+                for remote in role_matches.remotes.keys():
+                    hosts.add(remote)
+            elif isinstance(host_spec, basestring):
+                for remote in self.ctx.cluster.remotes.keys():
+                    if remote.name.split('@')[-1] == host_spec or \
+                            remote.shortname == host_spec:
+                        hosts.add(remote)
+        if not hosts:
+            raise RuntimeError("All target hosts were excluded!")
+        self.hosts = list(hosts)
+        hostnames = [h.shortname for h in self.hosts]
+        self.log.debug("Restricting task {name} to hosts: {hosts}".format(
+            name=self.name, hosts=' '.join(hostnames))
+        )
+        return self.hosts
+
+    def setup(self):
```

Comment:
Not really, it's just easy to forget to do.  Why make users of this base class remember to do that when they don't have to for begin, end or teardown?

Also, when I think of a setup method I want to compare it to setup methods in pytest or unittest where I don't need to ever call super().